File: lib/graph-service.ts
import core = require('@aws-cdk/core');
import lambda = require('@aws-cdk/aws-lambda');
import apigateway = require('@aws-cdk/aws-apigateway');

export class GraphService extends core.Construct {
    constructor(scope: core.Construct, id: string) {
        super(scope, id);

        const lambdas = new lambda.AssetCode('lambdas');

        const getAllGraphsHandler = new lambda.Function(this, 'GetAllGraphsHandler', {
            runtime: lambda.Runtime.NODEJS_10_X,
            code: lambdas,
            handler: 'graphs.getAll'
        });

        const addGraphHandler = new lambda.Function(this, 'AddGraphHandler', {
            runtime: lambda.Runtime.NODEJS_10_X,
            code: lambdas,
            handler: 'graphs.add'
        });

        const deleteGraphHandler = new lambda.Function(this, 'DeleteGraphHandler', {
            runtime: lambda.Runtime.NODEJS_10_X,
            code: lambdas,
            handler: 'graphs.delete'
        });

        const api = new apigateway.RestApi(this, 'graph-api', {
            restApiName: 'Graph Service',
            description: 'This service allows you to control graphs'
        });

        const getAllGraphsIntegration = new apigateway.LambdaIntegration(getAllGraphsHandler, {});
        const addGraphIntegration = new apigateway.LambdaIntegration(addGraphHandler, {});
        const deleteGraphIntegration = new apigateway.LambdaIntegration(deleteGraphHandler, {});

        api.root.addMethod('GET', getAllGraphsIntegration);
        api.root.addMethod('POST', addGraphIntegration);
        const graphEndpoint = api.root.addResource("{graphUrl}");
        graphEndpoint.addMethod('DELETE', deleteGraphIntegration);
    }
}

File: src/main/java/uk/gov/gchq/kai/service/GraphService.java
package uk.gov.gchq.kai.service;

import org.jetbrains.annotations.NotNull;

import software.amazon.awscdk.core.Construct;
import software.amazon.awscdk.services.apigateway.LambdaIntegration;
import software.amazon.awscdk.services.apigateway.Resource;
import software.amazon.awscdk.services.apigateway.RestApi;
import software.amazon.awscdk.services.lambda.AssetCode;
import software.amazon.awscdk.services.lambda.Code;
import software.amazon.awscdk.services.lambda.Function;
import software.amazon.awscdk.services.lambda.Runtime;

public class GraphService extends Construct {
    private static final AssetCode LAMBDAS = Code.fromAsset("lambdas");

    public GraphService(@NotNull final Construct scope, @NotNull final String id) {
        super(scope, id);

        Function getAllHandler = createFunction("getAllHandler", "graphs.getAll");
        Function deleteHandler = createFunction("deleteHandler", "graphs.delete");
        Function addHandler = createFunction("addHandler", "graphs.add");

        RestApi restApi = RestApi.Builder.create(this, "Kai-API").restApiName("Graph Service")
                .description("This service lets you interact with graphs").build();

        Resource graphResource = restApi.getRoot().addResource("graphs");

        graphResource.addMethod("GET", new LambdaIntegration(getAllHandler));
        graphResource.addMethod("POST", new LambdaIntegration(addHandler));
        graphResource.addResource("{graphId}").addMethod("DELETE", new LambdaIntegration(deleteHandler));
    }

    private Function createFunction(final String name, final String handler) {
        return Function.Builder.create(this, name).runtime(Runtime.NODEJS_10_X).code(GraphService.LAMBDAS)
                .handler(handler).build();
    }

}